[Enhancement] Improvements on Reactions

lschlesinger opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
I had to fix a bug that was found during testing, where a @MerlinMehmed  could add a reaction twice by selecting the emoji with some double clicks - which I could not reproduce.
As I checked the related code, I did some minor improvements preventing this unintended behavior and changed the current implementation w.r.t how/if the creation date of a post/answer post/reaction was built and persisted.
Added afterwards: During testing a saw that for the page discussion section, when loading took a while, I saw "Loading" and "No posts found" at the same time. I had to adapt the html file to show "No posts found" only after loading was finished. I forgot that in a previous PR.

### Description
<!-- Describe your changes in detail -->
Client Improvements:
- Reaction selector overlay is closed whenever the users clicks somewhere outside (previously one had to click on the emoji button that opened the selector)
- The emoji button that opened the selector has a different styling (active) when the selector overlay is open
- The client does not provide creation dates for posts and answer posts as these are added on the server-side

Server Improvements:
- Reactions creations date where not created/persisted, now they are
- Creation dates for posts and answer posts are created at server-side
- We now have server-side checks that a reaction can not be duplicated, i.e. a user cannot post the same reaction twice, most importantly, cannot vote twice

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis as any user
2. Navigate to a course with posts (testserver 1 has a course with many posts, its id is 477, you could use e.g. test user up to id 15)
3. create a post, no errors should occur, created date should be set correctly
4. create an answer post, no errors should occur, created date should be set correctly
5. edit a post, no errors should occur, creation date should be the same
6. edit an answer post, no errors should occur, creation date should be the same
7. react on a post and answer post, click on the reaction again -> should be removed; also very quickly double clicking/selecting a reaction should not lead to creating a reaction twice as a use

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

- Code Review
  - [x] Review 1
  - [x] Review 2
- Manual Tests
  - [x] Test 1
  - [x] Test 2

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->
One test for checking that duplicated reactions are prevented was added, existing tests are were adapted slightly
ReactionResource and ReactionService have 100% instruction coverage.

ndricimrr commented on src/main/java/de/tum/in/www1/artemis/service/metis/ReactionService.java:
```diff
@@ -78,6 +87,10 @@ public Reaction createReaction(Long courseId, Reaction reaction) {
             reaction.setAnswerPost(answerPost);
             reaction.setUser(user);
+            // check if the user already reacted with the reaction to add (-> avoid duplicated reactions) before saving
+            if (reactionRepository.findReactionsByAnswerPostIdAndEmojiIdAndUserId(answerPost.getId(), reaction.getEmojiId(), user.getId()).size() > 0) {
+                throw new BadRequestAlertException("You can only react once with a certain emoji", METIS_REACTION_ENTITY_NAME, "duplicatedReaction");
+            }
```
This seems like a repeated throw clause. Maybe it can be converted to a **orElseThrow** function like we usually do, since I've seen [them ](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server/?highlight=orelsethrow#avoid-service-dependencies)to be way more preferred. 
_ReactionRepository.java_
  ```
  default List<Reaction> findReactionsByPostIdAndEmojiIdAndUserIdElseThrow(Long postId, String emojiId, Long userId) throws EntityNotFoundException {
        return findReactionsByPostIdAndEmojiIdAndUserId(postId, emojiId, userId).stream().findAny().orElseThrow((() -> new EntityNotFoundException("Reactions", postId));
    }
```

then 

```suggestion
            reactionRepository.findReactionsByPostIdAndEmojiIdAndUserIdElseThrow(answerPost.getId(), reaction.getEmojiId(), user.getId())
```

lschlesinger replied:
The thing is, that it should throw an error _if_ there is an entity found, so the reaction would be duplicated (same user, same posting, same reaction). What you are suggesting is the other way around, throw an error if we cannot find the reaction we are searching for.

ndricimrr replied:
ah I see, then sth like this maybe, but that's up to you if you find it useful for your case or not.
```
    default void findNoReactionsByPostIdAndEmojiIdAndUserIdElseThrow(Long postId, String emojiId, Long userId) throws BadRequestAlertException {
        findReactionsByPostIdAndEmojiIdAndUserId(postId, emojiId, userId).stream().findAny().ifPresent(s -> { throw new BadRequestAlertException("You can only react once with a certain emoji", emojiId, "duplicatedReaction");});
    }
```

lschlesinger replied:
yep, that worked, I did it the way you proposed

ndricimrr commented on src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:
```diff
@@ -63,6 +64,10 @@ public Post createPost(Long courseId, Post post) {
         post.setAuthor(user);
+
+        // set creation date to now
+        post.setCreationDate(ZonedDateTime.now());
```
I see this added in multiple places. Since the date is now(). What if we were to set the date automatically upon entity **save()** like we do [here](url) and [here](https://github.com/ls1intum/Artemis/blob/develop/src/main/java/de/tum/in/www1/artemis/domain/analytics/TextAssessmentEvent.java#L33). 
For example in _Posting.java_ we could have something like this: 
```
    @Column(name = "creation_date")
    private ZonedDateTime creationDate = ZonedDateTime.now();
```
then when you do `postRepository.save(post);` on this entity, date is set automatically.

lschlesinger replied:
I changed it, thanks for the suggestion

ndricimrr reviewed (commented):
Looks good 👍  I added some suggestions for code improvements.

lschlesinger commented on src/main/java/de/tum/in/www1/artemis/service/metis/ReactionService.java:
```diff
@@ -78,6 +87,10 @@ public Reaction createReaction(Long courseId, Reaction reaction) {
             AnswerPost answerPost = answerPostService.findById(posting.getId());
             reaction.setAnswerPost(answerPost);
             reaction.setUser(user);
```
yes correct, good catch, I deleted 2 and 3

MerlinMehmed reviewed (commented):
It looks like the bug is fixed and I cannot reproduce it but it looks like that the translation is missing for the error message:
![Screenshot from 2021-09-19 18-16-54](https://user-images.githubusercontent.com/20031592/133932961-97e5a8c9-cf48-494b-9372-e162efd84cb5.png)

Olimbia23 reviewed (commented):
Checked the code, it looks good @lschlesinger 👍 
I agree with @ndricimrr suggestions.
I will approve after you have incorporated them.

lschlesinger commented:
@MerlinMehmed @Olimbia23 @ndricimrr thanks for testing! I incorporated all the changes ;)

ndricimrr reviewed (commented):
Code changes look good to me. Thanks for applying suggestions. 
Also Tested mentioned functionality, and it worked well.

For some unknown reason as you mentioned I can still trick the sytem to add multiple emojis. 

For example in course 477 > Discussion > Random I was able to add 23 reactions of one single type: 
![image](https://user-images.githubusercontent.com/18335489/134063128-9027b948-b408-4ca0-9bbb-2ae709a3ae29.png)

I clicked multiple times to make that work, but obviously very hard to reproduce manually again, so I'm thinking a rare race condition. However, I believe that one is able to reproduce this easily automatically by simply repeating the POST request. 
For example:
<details>

https://user-images.githubusercontent.com/18335489/134066877-2239ebb6-5b5a-41ef-aba3-e9124ede7398.mp4

</details>
Maybe the above video can help you reproduce locally and find where the issue is. Otherwise, if it's too rare of a race condition not worth the effort, I would also approve.

lschlesinger commented:
> @ndricimrr For some unknown reason as you mentioned I can still trick the sytem to add multiple emojis.

whaaaat 🤯 ? that is really evil 😄

lschlesinger commented:
@ndricimrr I think I found a nice solution for the duplication problem that does not rely on querying the db for existing reactions. I added a constraint on the table that ensures uniqueness w.r.t. user_id, emoji_id and post_id or answer_post_id. As we might already have some of those duplicates in our databases (in production and on test servers) I added an sql script that deletes those duplicates and only keeps one of them. Then the two constraints are added and we prevent new duplicates being added to the db. I tested it with your approach of sending many requests at the same time, for every duplicate reaction you will get an internal server error and the reaction is not added to the db. Let me know, what you think about this approach :) and thanks again for bringing this issue up!

Olimbia23 reviewed (approved):
Nice work @lschlesinger.
Thank you for incorporating the suggestions and fixing the race condition issue.
Code changes look good. Approve 👍

ndricimrr reviewed (approved):
Thanks for applying suggestions. Changes look good. Tested and working as expected

krusche commented on src/main/resources/config/liquibase/changelog/20210922211100_changelog.xml:
```diff
@@ -0,0 +1,16 @@
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.9.xsd">
+    <changeSet author="schlesinger" id="20210922211100">
+        <sql>
+            DELETE FROM reaction
```
please add some natural language comments what this sql query will do

krusche replied:
and a better formatting would be nice (inner queries should e.g. be indented)

lschlesinger replied:
done in 3e661c3

MerlinMehmed reviewed (dismissed):
The bug was fixed, I cannot react several times with one reaction anymore. Good job!

argertboja reviewed (approved):
Tested on TS1 and I can only vote once as expected 🙂 Nice job @lschlesinger 👍🏼

MerlinMehmed reviewed (approved):
Reapprove! Retested on TS1, still works.
